go/storage/mkvs/db/pathbadger/node.go

```diff
@@ -0,0 +1,531 @@
+			}
+			pos += size
+			n.Right = ptr
+		}
+
+		// Optional value of leaf node if there is anything left.
+		if len(value) > pos {
+			leaf := node.LeafNode{
+				Clean: true,
+			}
+			size, err := leaf.Key.SizedUnmarshalBinary(value[pos:])
+			if err != nil {
+				return nil, fmt.Errorf("failed to unmarshal leaf node size: %w", err)
+			}
+			pos += size
+
+			leaf.Value = append([]byte{}, value[pos:]...) // Copy value.
+			leaf.UpdateHash()
+
+			n.LeafNode = &node.Pointer{
+				Clean:      true,
+				Hash:       leaf.Hash,
+				Node:       &leaf,
+				DBInternal: newInvalidDbPtr(), // Not standalone.
+			}
+		}
+
+		n.UpdateHash()
+
+		return &n, nil
+	default:
+		return nil, fmt.Errorf("mkvs: unsupported node kind '%X' (db corruption?)", kind)
+	}
+}
+
+// ptrToDb serializes a MKVS pointer into an internal database representation.
+func ptrToDb(ptr *node.Pointer) []byte {
+	iptr := ptr.DBInternal.(*dbPtr)
+	if iptr.isInvalid() {
+		panic("mkvs/pathbadger: attempted to serialize invalid internal pointer")
+	}
+
+	h, _ := ptr.Hash.MarshalBinary()
+	data := append([]byte{}, h...)
+	data = append(data, encodeNodeKey(iptr.version, iptr.index)...)
+
+	return data
+}
+
+func encodeVersionKey(version uint64) []byte {
+	var rawVersion [8]byte
+	binary.BigEndian.PutUint64(rawVersion[:], version)
+	return rawVersion[:]
+}
+
+func encodeIndexKey(index uint32) []byte {
+	var rawIndex [4]byte
+	binary.BigEndian.PutUint32(rawIndex[:], index)
+	return rawIndex[:]
+}
+
+func encodeNodeKey(version uint64, index uint32) []byte {
+	data := append([]byte{}, encodeVersionKey(version)...)
+	data = append(data, encodeIndexKey(index)...)
+	return data
+}
+
+// dbPtr contains internal metadata needed for pointer resolution.
+type dbPtr struct {
+	version uint64
+	index   uint32
+}
+
+func newInvalidDbPtr() *dbPtr {
+	return &dbPtr{
+		version: versionInvalid,
+		index:   indexInvalid,
+	}
+}
+
+func (p *dbPtr) isRoot() bool {
+	return p.index == indexRootNode
+}
+
+func (p *dbPtr) isInvalid() bool {
+	return p.version == versionInvalid && p.index == indexInvalid
+}
+
+func (p *dbPtr) dbKey() []byte {
+	return encodeNodeKey(p.version, p.index)
+}
+
+// Implements node.DBPointer.
+func (p *dbPtr) SetDirty() {
+	p.version = versionInvalid
+	p.index = indexInvalid
+}
+
+// Implements node.DBPointer.
+func (p *dbPtr) Clone() node.DBPointer {
+	return &dbPtr{
+		version: p.version,
+		index:   p.index,
+	}
+}
+
+// ptrFromDb deserializes a pointer from internal database representation.
+func ptrFromDb(data []byte) (int, *node.Pointer, error) {
+	// Format: [hash] [version] [index]
+
+	// Validate data size.
+	if len(data) < hash.Size+8+4 {
+		return -1, nil, fmt.Errorf("malformed pointer (not enough bytes)")
+	}
+
+	var (
+		h   hash.Hash
+		err error
+	)
+	if err = h.UnmarshalBinary(data[:hash.Size]); err != nil {
+		return -1, nil, fmt.Errorf("failed to unmarshal hash: %w", err)
+	}
+	if h.IsEmpty() {
+		// Empty hashes are not allowed in serialized form.
```

Comment:
Because a pointer to an empty node should not even be present in the database? E.g. an internal node with only the left pointer and an empty right pointer should be serialized as `kindInternalWithLeft` where the right pointer is not present. Or if the internal node doesn't have a leaf node, the leaf node should not be serialized.

We can add a panic during serialization.

Reply:
Its fine if in a round-trip test serialization panics, but its not the best if serialization succeeds and deserialization panics. So I would panic during serialization also.